### DISCUSSION TITLE
`@ApplicationScoped` and `@Scheduled`  with `@UnlessBuildProperty`

### DISCUSSION THREAD

**Discussion by kirankolli-addepar** [quarkusio/quarkus#39994] [Q&A] [answered]
I have a service class like so : 

```
@UnlessBuildProperty(name = "someservice.enabled", stringValue = "false")
@ApplicationScoped
public class SomeService{

  @Scheduled(every = "${check.interval:1M}")
  public void process() {
     //do something
  }
}
```

The intention is that the service class should not be instantiated at all in case the build property `someservice.enabled` is `false` . And if not specified or `true` the service should be instantiated and the scheduled method in the service should execute every 1 minute. 

However with the `@UnlessBuildProperty` either the class is never instantiated or the scheduled method never executes. 

Only when I comment out the above annotation does it work. What am I doing wrong ? And how should I fix this ?

[a comment by an automated account is not shown]

**Comment by manovotn** [comment, marked as answer]
> And if not specified...

For this you need to specify the `enableIfMissing` property on the annotation, see [this](https://github.com/quarkusio/quarkus/blob/main/extensions/arc/runtime/src/main/java/io/quarkus/arc/properties/UnlessBuildProperty.java#L37).

>  ... the service should execute every 1 minute.

Are you sure the expression you are using is evaluating correctly? I.e. `@Scheduled(every = "${check.interval:1M}")` 
For an execution once a min, I think you could use something like `@Scheduled(every="1m")`.

**Reply by kirankolli-addepar** [reply]
@manovotn on the `Scheduled` annotation thats a way I use to parameterise the check interval. Allowing me to override with a property or java env variable if needed.

**Reply by manovotn** [reply]
@kirankolli-addepar I did a quick local test with a class looking like this:

```
@UnlessBuildProperty(name = "someservice.enabled", stringValue = "false", enableIfMissing = true)
@ApplicationScoped
public class ScheduledBean {


    @Scheduled(every = "10s")
    public void process() {
        System.err.println("~~~~~~~~~~~ SOMETHING");
    }
}
```

And this works as expected - the bean is enabled and triggering the task so long as `someservice.enabled` isn't `false`.
Note that I added `enableIfMissing = true` which makes sure the bean is enabled if the property is missing entirely.

So if this doesn't work for you, a reproducer would be useful otherwise it's difficult to say where the problem lies :)

**Reply by kirankolli-addepar** [reply]
@manovotn , the weird part was I did have a config class like so : 

```
@ConfigMapping(prefix = "someservice")
public interface SomeServiceConfig {
 @WithName("enabled")
  @WithDefault("true")
  Optional<Boolean> enabled();
}
```

so I expected that changing this value would work in conjunction with the change above. However this was not working until I explicitly added a property in `application.properties` .

**Reply by manovotn** [reply]
>   @WithDefault("true")
>  Optional<Boolean> enabled();

What's the use case for a combination of `Optional` and `@WithDefault`? It seems like you wanted to either have it optional and not having default is fine or you want to always have some value in which case you don't use optional and set a default. Or am I missing something?

**Reply by manovotn** [reply]
@kirankolli-addepar FYI after some more digging and testing I agree that the behavior with `@ConfigMapping` is weird at best.
I have created the following issue to track it - https://github.com/quarkusio/quarkus/issues/40034
